packages/powersync_attachments_stream/lib/src/attachment_queue_service.dart

```diff
@@ -0,0 +1,392 @@
+      attachmentsService: attachmentsService,
+      errorHandler: errorHandler,
+      syncThrottle: syncThrottleDuration,
+      period: syncInterval,
+    );
+  }
+
+  /// Initialize the attachment queue by:
+  /// 1. Creating the attachments directory.
+  /// 2. Adding watches for uploads, downloads, and deletes.
+  /// 3. Adding a trigger to run uploads, downloads, and deletes when the device is online after being offline.
+  Future<void> startSync() async {
+    await _mutex.lock(() async {
+      if (_closed) {
+        throw Exception('Attachment queue has been closed');
+      }
+
+      await _stopSyncingInternal();
+
+      await localStorage.initialize();
+
+      await attachmentsService.withContext((context) async {
+        await _verifyAttachments(context);
+      });
+
+      await syncingService.startSync();
+
+      // Listen for connectivity changes and watched attachments
+      _syncStatusSubscription = watchAttachments().listen((items) async {
+        await _processWatchedAttachments(items);
+      });
+
+      logger.info('AttachmentQueue started syncing.');
+    });
+  }
+
+  /// Stops syncing. Syncing may be resumed with [startSync].
+  Future<void> stopSyncing() async {
+    await _mutex.lock(() async {
+      await _stopSyncingInternal();
+    });
+  }
+
+  Future<void> _stopSyncingInternal() async {
+    if (_closed) return;
+
+    await _syncStatusSubscription?.cancel();
+    _syncStatusSubscription = null;
+    await syncingService.stopSync();
+
+    logger.info('AttachmentQueue stopped syncing.');
+  }
+
+  /// Closes the queue. The queue cannot be used after closing.
+  Future<void> close() async {
+    await _mutex.lock(() async {
+      if (_closed) return;
+
+      await _syncStatusSubscription?.cancel();
+      await syncingService.close();
+      _closed = true;
+
+      logger.info('AttachmentQueue closed.');
+    });
+  }
+
+  /// Resolves the filename for new attachment items.
+  /// Concatenates the attachment ID and extension by default.
+  Future<String> resolveNewAttachmentFilename(
+    String attachmentId,
+    String? fileExtension,
+  ) async {
+    return fileExtension != null
+        ? '$attachmentId.$fileExtension'
+        : '$attachmentId.dat';
+  }
+
+  /// Processes attachment items returned from [watchAttachments].
+  /// The default implementation asserts the items returned from [watchAttachments] as the definitive
+  /// state for local attachments.
+  Future<void> _processWatchedAttachments(
+    List<WatchedAttachmentItem> items,
+  ) async {
+    await attachmentsService.withContext((context) async {
+      final currentAttachments = await context.getAttachments();
+      final List<Attachment> attachmentUpdates = [];
+
+      for (final item in items) {
+        final existingQueueItem = currentAttachments
+            .cast<Attachment?>()
+            .firstWhere(
+              (a) => a != null && a.id == item.id,
+              orElse: () => null,
+            );
```

Comment:
I think a more intuitive way to express this without the `cast()` may be
```suggestion
            .where((a) => a.id == item.id)
            .firstOrNull;
```